src/graph/LookupExecutor.cpp

```diff
@@ -240,77 +241,212 @@ Status LookupExecutor::checkFilter() {
+            return f.get_type().get_type() == nebula::cpp2::SupportedType::STRING;
+        });
+        if (stringField != fields.end()) {
+            continue;
+        }
+        for (const auto& filter : filters_) {
+            auto it = std::find_if(fields.begin(), fields.end(),
+                                   [filter](const auto &field) {
+                                       return field.get_name() == filter.first;
+                                   });
+            if (it == fields.end()) {
+                allColsHint = false;
                 break;
             }
         }
-        if (!matching || index->get_fields().size() < filterCols.size()) {
-            continue;
+        if (allColsHint) {
+            validIndexes.emplace_back(index);
         }
-        indexes.emplace_back(index);
     }
+    // If the first field of the index does not match any condition, the index is invalid.
+    // remove it from validIndexes.
+    if (!validIndexes.empty()) {
+        auto index = validIndexes.begin();
+        while (index != validIndexes.end()) {
+            const auto& fields = index->get()->get_fields();
+            auto it = std::find_if(filters_.begin(), filters_.end(),
+                                   [fields](const auto &filter) {
+                                       return filter.first == fields[0].get_name();
+                                   });
+            if (it == filters_.end()) {
+                validIndexes.erase(index);
+            } else {
+                index++;
+            }
+        }
+    }
+    return validIndexes;
+}
 
-    if (indexes.empty()) {
-        return Status::IndexNotFound();
+std::vector<std::shared_ptr<nebula::cpp2::IndexItem>> LookupExecutor::findValidIndex() {
+    // Check contains string type field from where condition
+    auto *sm = ectx()->schemaManager();
+    auto schema = isEdge_
+                  ? sm->getEdgeSchema(spaceId_, tagOrEdge_)
+                  : sm->getTagSchema(spaceId_, tagOrEdge_);
+    if (schema == nullptr) {
+        LOG(ERROR) << "No schema found : " << from_;
+        return {};
     }
 
-    /**
-     * step 2 , if have multiple valid indexes, get the best one.
-     * for example : if where clause is :
-     * col1 > 1 and col2 > 1 --> index1 and index2 are valid. get one of these at random.
-     * col1 > 1 and col2 == 1 --> index1 and index2 are valid.
-     *                            but need choose one for storage layer.
-     *                            here index2 is chosen because col2 have a equivalent value.
-     */
-    std::map<int32_t, IndexID> indexHint;
+    bool hasStringCol = false;
+    for (const auto& filter : filters_) {
+        auto type = schema->getFieldType(filter.first);
+        if (type.get_type() == nebula::cpp2::SupportedType::STRING) {
+            hasStringCol = true;
+            break;
+        }
+    }
+    return hasStringCol ? findValidIndexWithStr() : findValidIndexNoStr();
```

Comment:
Add more comments about each method. I can't figure out what your logic in your code.